### PR TITLE
Autoplay

### PR DESCRIPTION
**Pull request by erwinmombay**

### CODE REVIEW

**Comment by erwinmombay** [comment]
relies on https://github.com/ampproject/amphtml/pull/555 and will wait for that to be merged in until i ask this to be reviewed.

File: extensions/amp-carousel/0.1/slides.js [hunk 1]
```diff
@@ -78,6 +98,50 @@ export class AmpSlides extends BaseCarousel {
...
+   */
+  setupAutoplay_() {
+    if (!this.isAutoplaying_) {
+      return;
+    }
+
+    let autoplayValue = Number(this.element.getAttribute('autoplay'));
+    // If it isn't a number and is not greater than 0 then don't assign
+    // and use the default.
+    if (!Number.isNaN(autoplayValue) && autoplayValue > 0) {
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: Should just be able to check for positive `autoplayValue`:

```js
NaN > 0; // => false
```

File: extensions/amp-carousel/0.1/slides.js [hunk 2]
```diff
@@ -45,6 +55,16 @@ export class AmpSlides extends BaseCarousel {
...
+
+    /** @private {?number} */
+    this.autoplayTimeoutId_ = null;
+
+    this.setupAutoplay_();
+
+    // Setup microtask to let the runtime settle.
+    // TODO: (erwinm) experiment with only letting autoplay run if the
+    // carouse is in the viewport.
+    timer.delay(this.tryAutoplay_.bind(this , 1, true), 0);
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: How about setting the timer in `#setupAutoplay_`? It'll avoid setting up this timer when there's no `autoplay` attribute, and guard against non-positive values.

**Comment by dvoytenko** [comment]
Still WIP or ready to review?

**Comment by erwinmombay** [comment]
@dvoytenko still WIP cause of the other, PR (loop feature). missing a feature there

File: extensions/amp-carousel/0.1/slides.js (hunk 1 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: thanks :+1: , will update

File: extensions/amp-carousel/0.1/slides.js (hunk 2 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: good point, moving

Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: done.

File: extensions/amp-carousel/0.1/slides.js (hunk 1 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: done.

**Comment by erwinmombay** [comment]
@dvoytenko this is ready for review

File: extensions/amp-carousel/0.1/slides.js [hunk 3]
```diff
@@ -81,6 +93,55 @@ export class AmpSlides extends BaseCarousel {
...
+      return;
+    }
+
+    let autoplayValue = Number(this.element.getAttribute('autoplay'));
+    // If it isn't a number and is not greater than 0 then don't assign
+    // and use the default.
+    if (autoplayValue > 0) {
+      // Guard against autoplayValue that is lower than 1s to prevent
+      // people from crashing the runtime with providing very low delays.
+      this.autoplayDelay_ = autoplayValue < 1000 ? 1000 : autoplayValue;
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: How about `Math.max(1000, autoplayValue)`?

File: extensions/amp-carousel/0.1/slides.js [hunk 4]
```diff
@@ -81,6 +93,55 @@ export class AmpSlides extends BaseCarousel {
...
+   */
+  tryAutoplay_(dir, animate) {
+    if (!this.isAutoplaying_) {
+      return;
+    }
+
+    if (this.autoplayTimeoutId_ !== null) {
+      timer.cancel(this.autoplayTimeoutId_);
+    }
+    this.autoplayTimeoutId_ = timer.delay(this.go.bind(this, 1, animate),
```
Comment (jridgewell) [inline extensions/amp-carousel/0.1/slides.js]: `this.go.bind(this, dir, animate)`?

File: extensions/amp-carousel/0.1/slides.js (hunk 3 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: yep, better.

done.

File: extensions/amp-carousel/0.1/slides.js (hunk 4 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: thanks for catching!

done.

File: extensions/amp-carousel/amp-carousel.md [hunk 5]
```diff
@@ -53,6 +53,14 @@ unless only a single child is present.
 
 If present, the user may advance past the first item or the final item.
 
+**autoplay** (type=slides only)
+
+If present, advances the slide to the next slide without user interaction.
+By default it will advance in 5000 milliseconds (5 seconds) but will use the
```
Comment (AaronCritchley) [inline extensions/amp-carousel/amp-carousel.md]: Consider '5000 millisecond intervals' as a more readable alternative :smile:

Comment (erwinmombay) [inline extensions/amp-carousel/amp-carousel.md]: that definitely is better. will change

File: extensions/amp-carousel/0.1/slides.js [hunk 6]
```diff
@@ -81,6 +93,55 @@ export class AmpSlides extends BaseCarousel {
...
+    // and use the default.
+    if (autoplayValue > 0) {
+      // Guard against autoplayValue that is lower than 1s to prevent
+      // people from crashing the runtime with providing very low delays.
+      this.autoplayDelay_ = Math.max(1000, autoplayValue);
+    }
+
+    // By default `autoplay` should also mean that the current carousel slide
+    // is looping. (to be able to advance past the last item)
+    if (!this.element.hasAttribute('loop')) {
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slides.js]: Do you not need to update this.isLooping_ var?

File: extensions/amp-carousel/0.1/slides.js [hunk 7]
```diff
@@ -81,6 +93,55 @@ export class AmpSlides extends BaseCarousel {
...
+      this.element.setAttribute('loop', '');
+    }
+
+    // Setup microtask to let the runtime settle.
+    // TODO: (erwinm) experiment with only letting autoplay run if the
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slides.js]: I think this is a must do.

Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: got it, will implement.

File: extensions/amp-carousel/0.1/slides.js [hunk 8]
```diff
@@ -37,6 +38,12 @@ export class AmpSlides extends BaseCarousel {
     /** @private @const {boolean} */
     this.isLooping_ = this.element.hasAttribute('loop');
 
+    /** @private @const {boolean} */
+    this.isAutoplaying_ = this.element.hasAttribute('autoplay');
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slides.js]: What about isAutoplayRequested? "autoplaying" sounds like playing right now.

File: extensions/amp-carousel/0.1/slides.js (hunk 6 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: yes, should update. fixing

File: extensions/amp-carousel/0.1/slides.js [hunk 9]
```diff
@@ -81,6 +93,55 @@ export class AmpSlides extends BaseCarousel {
...
+   */
+  tryAutoplay_(dir, animate) {
+    if (!this.isAutoplaying_) {
+      return;
+    }
+
+    if (this.autoplayTimeoutId_ !== null) {
+      timer.cancel(this.autoplayTimeoutId_);
+    }
+    this.autoplayTimeoutId_ = timer.delay(this.go.bind(this, dir, animate),
```
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slides.js]: So what would happen if the ready slides manually? I think we either need to restart autoplay or cancel it entirely. I'm leaning toward the later.

Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: do you mean if the user interactions and advances it manually? if so then we restart autoplay, which seems to be pretty common pattern in carousel components.

File: extensions/amp-carousel/0.1/slides.js (hunk 7 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: @dvoytenko are you ok with with me doing it in a separate PR?

Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slides.js]: If fast-follow and I have a bug# - then yes.

File: extensions/amp-carousel/0.1/slides.js (hunk 9 above)
Comment (dvoytenko) [inline extensions/amp-carousel/0.1/slides.js]: Yeah, we can just continue autoplay for now. I think this is easy to address later.

File: extensions/amp-carousel/0.1/slides.js (hunk 7 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: yeah i should get to it tonight. issue => https://github.com/ampproject/amphtml/issues/605

File: extensions/amp-carousel/amp-carousel.md (hunk 5 above)
Comment (erwinmombay) [inline extensions/amp-carousel/amp-carousel.md]: done.

File: extensions/amp-carousel/0.1/slides.js (hunk 9 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: got it, will file a bug for research and implementation if needed.

File: extensions/amp-carousel/0.1/slides.js (hunk 6 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: done.

File: extensions/amp-carousel/0.1/slides.js (hunk 8 above)
Comment (erwinmombay) [inline extensions/amp-carousel/0.1/slides.js]: done.

**Comment by erwinmombay** [comment]
research task added: https://github.com/ampproject/amphtml/issues/606
bug followup added: https://github.com/ampproject/amphtml/issues/605

@dvoytenko PTAL

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**